File: src/main/java/com/wafflestudio/draft/security/password/UserPrincipal.java
package com.wafflestudio.draft.security.password;

import com.wafflestudio.draft.model.User;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class UserPrincipal extends User implements UserDetails {

    public UserPrincipal(User user) {
        super(user.getUsername(), user.getEmail());
        super.setId(user.getId());
        super.setPassword(user.getPassword());
        super.setRoles(user.getRoles());
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        if (super.getRoles() == null) return new ArrayList<>();

        List<String> roles = Arrays.asList(super.getRoles().split(","));
        return roles.stream().map(SimpleGrantedAuthority::new).collect(Collectors.toList());
    }


    @Override
    public String getPassword() {
        return super.getPassword();
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/kotlin/com/wafflestudio/draft/security/password/UserPrincipal.kt
package com.wafflestudio.draft.security.password

import com.wafflestudio.draft.model.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.util.*
import java.util.stream.Collectors

class UserPrincipal(val user: User) : UserDetails {
    override fun getUsername(): String {
        return user.username
    }

    override fun getPassword(): String? {
        return user.password
    }

    override fun getAuthorities(): Collection<GrantedAuthority> {
        if (user.roles == null) return ArrayList()
        val roles = Arrays.asList(*user.roles!!.split(",").toTypedArray())
        return roles.stream().map { role: String? -> SimpleGrantedAuthority(role) }.collect(Collectors.toList())
    }


    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return true
    }


}
